src/appengine/libs/auth.py

```diff
@@ -54,6 +59,66 @@ def is_current_user_admin():
   return bool(key.get())
 
 
+@memoize.wrap(memoize.FifoInMemory(1))
+def _project_number_from_id(project_id):
+  """Get the project number from project ID."""
+  resource_manager = build('cloudresourcemanager', 'v1')
+  result = resource_manager.projects().get(projectId=project_id).execute()
+  return result['projectNumber']
+
+
+@memoize.wrap(memoize.FifoInMemory(1))
+def _get_iap_key(key_id):
+  """Retrieves a public key from the list published by Identity-Aware Proxy,
+  re-fetching the key file if necessary.
+  """
+  resp = requests.get('https://www.gstatic.com/iap/verify/public_key')
+  if resp.status_code != 200:
+    raise AuthError('Unable to fetch IAP keys: {} / {} / {}'.format(
+        resp.status_code, resp.headers, resp.text))
+
+  result = resp.json()
+  key = result.get(key_id)
+  if not key:
+    raise AuthError('Key {!r} not found'.format(key_id))
+
+  return key
+
+
+def _validate_iap_jwt(iap_jwt):
+  """Validate JWT assertion."""
+  project_id = utils.get_application_id()
+  expected_audience = '/projects/{}/apps/{}'.format(
+      _project_number_from_id(project_id), project_id)
+
+  try:
+    key_id = jwt.get_unverified_header(iap_jwt).get('kid')
+    if not key_id:
+      raise AuthError('No key ID.')
+
+    key = _get_iap_key(key_id)
+    decoded_jwt = jwt.decode(
+        iap_jwt,
+        key,
+        algorithms=['ES256'],
+        issuer='https://cloud.google.com/iap',
+        audience=expected_audience)
+    return decoded_jwt['sub'], decoded_jwt['email']
+  except (jwt.exceptions.InvalidTokenError,
+          requests.exceptions.RequestException) as e:
+    raise AuthError('JWT assertion decode error ' + str(e))
+
+
+def get_iap_email(current_request):
+  """Get Cloud IAP email."""
+  jwt_assertion = current_request.headers.get('X-Goog-IAP-JWT-Assertion')
+  if not jwt_assertion:
+    return None
+
+  _, email = _validate_iap_jwt(jwt_assertion)
```

Comment:
Since we are only user of _validate_iap_jwt, we can skip returning first argument ?

Reply:
Done!